homie-controller/tests/integration.rs

```diff
@@ -0,0 +1,220 @@
+            }
+        }
+    }
+
+    // Check that the device looks how we expect.
+    let devices = controller.devices();
+    let device = devices.get("device_id").unwrap();
+    let node = device.nodes.get("node_id").unwrap();
+    let property = node.properties.get("property_id").unwrap();
+    log::info!("Property: {:?}", property);
+    assert_eq!(property.value(), Ok(42));
+
+    // Send a value from the controller to the device.
+    controller
+        .set("device_id", "node_id", "property_id", 13)
+        .await
+        .unwrap();
+
+    // Wait for the device to receive the value and send it back to the controller.
+    loop {
+        if let Some(event) = controller.poll(&mut event_loop).await.unwrap() {
+            log::trace!("Event: {:?}", event);
+            if let Event::PropertyValueChanged {
+                device_id,
+                node_id,
+                property_id,
+                value,
+                fresh,
+            } = event
+            {
+                assert_eq!(device_id, "device_id");
+                assert_eq!(node_id, "node_id");
+                assert_eq!(property_id, "property_id");
+                assert_eq!(value, "13");
+                assert_eq!(fresh, true);
+                break;
+            }
+        }
+    }
+    assert_eq!(*updates.lock().unwrap(), vec!["13".to_string()]);
+
+    // Check that the value sent back is reflected on the controller's view of the device.
+    let devices = controller.devices();
+    let device = devices.get("device_id").unwrap();
+    let node = device.nodes.get("node_id").unwrap();
+    let property = node.properties.get("property_id").unwrap();
+    log::info!("Property: {:?}", property);
+    assert_eq!(property.value(), Ok(13));
+
+    // Disconnect the device.
+    homie.disconnect().await.unwrap();
+    let err = homie_handle.await.unwrap_err();
+    if let SpawnError::Connection(ConnectionError::MqttState(StateError::Io(e))) = err {
+        assert_eq!(e.kind(), ErrorKind::ConnectionAborted);
+    } else {
+        panic!("Unexpected error {:?}", err);
+    }
+}
+
+/// Spawn an MQTT broker listening on the given port.
+fn spawn_mqtt_broker(port: u16) {
+    // TODO: Construct Config directly once fields are made public.
+    let broker_config = toml::from_str::<Config>(&format!(
+        r#"
+    id = 0
+
+    [router]
+    id = 0
+    dir = "/tmp/integration/rumqttd"
+    max_segment_size = 10240
+    max_segment_count = 10
+    max_connections = 20
+
+    [servers.1]
+    port = {}
+    next_connection_delay_ms = 1
+
+    [servers.1.connections]
+    connection_timeout_ms = 100
+    max_client_id_len = 100
+    throttle_delay_ms = 0
+    max_payload_size = 2048
+    max_inflight_count = 500
+    max_inflight_size = 1024
+
+    [console]
+    port = 0
+    "#,
+        port
+    ))
+    .unwrap();
+    let mut broker = Broker::new(broker_config);
+    thread::spawn(move || {
+        broker.start().unwrap();
+    });
+}
```

Comment:
fancy fixing the clippy problems below while you're here?
|
|
V

Reply:
Will do in a different PR.